homeassistant/auth/providers/openid.py

```diff
@@ -0,0 +1,282 @@
+"""OpenID based authentication provider."""
+
+import logging
+import re
+from secrets import token_hex
+from typing import Any, Dict, Optional, cast
+
+from aiohttp import ClientResponseError
+from aiohttp.client import ClientResponse
+import voluptuous as vol
+from yarl import URL
+
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.config_entry_oauth2_flow import (
+    AUTH_CALLBACK_PATH,
+    _encode_jwt,
+    async_register_view,
+)
+from homeassistant.helpers.network import get_url
+
+from . import AUTH_PROVIDER_SCHEMA, AUTH_PROVIDERS, AuthProvider, LoginFlow
+from ..models import Credentials, UserMeta
+
+REQUIREMENTS = ["python-jose==3.1.0"]
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_CLIENT_ID = "client_id"
+CONF_CLIENT_SECRET = "client_secret"
+CONF_CONFIGURATION = "configuration"
+CONF_EMAILS = "emails"
+CONF_SUBJECTS = "subjects"
+
+CONFIG_SCHEMA = AUTH_PROVIDER_SCHEMA.extend(
+    {
+        vol.Required(CONF_CONFIGURATION): str,
+        vol.Required(CONF_CLIENT_ID): str,
+        vol.Required(CONF_CLIENT_SECRET): str,
+        vol.Optional(CONF_EMAILS, default=[]): [str],
+        vol.Optional(CONF_SUBJECTS, default=[]): [str],
```

Comment:
Should these be exclusive? Like, it's not clear now that either one matching would grant access.

Also we should not use default as a list. Imagine this going wrong somewhere and someone accidentally changes these lists, impacting every OpenID implementation that is loaded.

Reply:
oh right. it's the "same" list every where. That damn python gotcha.